Ensure lint workflow checks whether 3rd party license and code is up to date

andyfeller opened the pull request:
fixes cli/cli#9422
fixes github/cli#111

This pull request addresses GitHub CLI obligation to comply with the licenses of our 3rd party dependencies.

These changes update the `lint` workflow to verify license information is up to date including source code certain licenses require redistributing source code around.

Additionally, this pull request introduces `scripts/license` and `scripts/license-check` to help maintainers and contributors regenerate this information easily.

andyfeller committed:
Adopt license compliance scripts into workflows, docs

This commit introduces the use of `go-licenses` within CI/CD and manual processes for generating / updating the license information used by GitHub CLI including the code required by license to be redistributed.

During GitHub CLI pull requests, the `lint` workflow will notify users if this information is not updated.

andyfeller committed:
Update 3rd party license information

BagToad commented on docs/license-compliance.md:
````diff
@@ -0,0 +1,52 @@
+   ```shell
+   git add third-party-licenses.*.md third-party/
+   git commit -m "Update third-party license information"
+   ```
````
_not blocking, thought_: might be nice for this to be a `make licenses` instead

andyfeller commented on third-party-licenses.windows.md:
```diff
@@ -0,0 +1,187 @@
+- [github.com/mailru/easyjson](https://pkg.go.dev/github.com/mailru/easyjson) ([MIT](https://github.com/mailru/easyjson/blob/v0.9.0/LICENSE))
+- [github.com/mattn/go-colorable](https://pkg.go.dev/github.com/mattn/go-colorable) ([MIT](https://github.com/mattn/go-colorable/blob/v0.1.14/LICENSE))
+- [github.com/mattn/go-isatty](https://pkg.go.dev/github.com/mattn/go-isatty) ([MIT](https://github.com/mattn/go-isatty/blob/v0.0.20/LICENSE))
+- [github.com/mattn/go-localereader](https://pkg.go.dev/github.com/mattn/go-localereader) ([Unknown](Unknown))
```
So this is an interesting call out because https://github.com/mattn/go-localereader/blob/master/LICENSE does in fact say it is under MIT license 🤔 

Definitely sounds like a follow up with https://github.com/google/go-licenses issue as we are not managing this by hand.

andyfeller replied:
xlinking https://github.com/google/go-licenses/issues/275#issuecomment-2922931852 for follow up

BagToad commented on script/licenses-check:
```diff
@@ -0,0 +1,21 @@
+    #       Normally these warnings are packages containing non go code, which may or may not require explicit attribution,
+    #       depending on the license.
+    echo "Checking licenses for ${goos}..."
+    GOOS="${goos}" go-licenses report ./... --template .github/licenses.tmpl --ignore github.com/cli/cli > third-party-licenses.${goos}.copy.md || echo "Ignore warnings"
```
question: should the output be stored in a tempdir just like we do with the generation?

andyfeller committed:
Avoid analyzing 3rd party license content with CodeQL

With these changes, `cli/cli` will be redistributing code as-is due to license compliance, which we will not change or address issues around.  Without these changes, our pull requests are getting a bunch of false positive annotations we cannot and will not fix directly.

andyfeller committed:
Merge branch 'trunk' into andyfeller/9422-license-compliance

andyfeller committed:
Merge branch 'trunk' into andyfeller/9422-license-compliance

andyfeller committed:
Update 3rd party licenses

andyfeller committed:
Use temp directory for license checks

andyfeller committed:
Use `make` for license generation and checks

vovong92 commented:
andyfeller/9422-license-compliance

BagToad commented on script/licenses:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+go install github.com/google/go-licenses@latest
```
Question: Are we comfortable with always using the latest here, or should we pin to a sha for stability/security etc?

I know that pinning comes with the "when does this get updated then?" question, and I don't know the answer to that either :P

BagToad replied:
I guess this is especially in my mind with this script being used in actions... 🤔

andyfeller replied:
@BagToad : what do you think about conditionalizing this to skip installing this if `CI`, leaving that to the workflow?

otherwise, everything running the script has to manage its installation.  conditionalizing it will make it easy for contributors and maintainers.

BagToad commented on script/licenses:
```diff
@@ -0,0 +1,25 @@
+export TEMPDIR="$(mktemp -d)"
+trap "rm -fr ${TEMPDIR}" EXIT
+
+# Clear third-party source code to avoid stale content
+rm -rf third-party
+mkdir -p third-party
+
```
Question: Should we check successes (exit codes) here? Are we okay with this maybe failing but then continuing?

andyfeller replied:
I'm not worried about these commands failing, but if there is a concern then we might want to `set -e` so any errors cause the whole script to fail.

BagToad commented on script/licenses:
```diff
@@ -0,0 +1,25 @@
+
+# Setup temporary directory to collect updated third-party source code
+export TEMPDIR="$(mktemp -d)"
+trap "rm -fr ${TEMPDIR}" EXIT
```
(praise) Smart!

BagToad commented on script/licenses:
```diff
@@ -0,0 +1,25 @@
+    echo "Generating licenses for ${goos}..."
+    GOOS="${goos}" go-licenses save ./... --save_path="${TEMPDIR}/${goos}" --force || echo "Ignore warnings"
+    GOOS="${goos}" go-licenses report ./... --template .github/licenses.tmpl --ignore github.com/cli/cli > third-party-licenses.${goos}.md || echo "Ignore warnings"
+    cp -fR "${TEMPDIR}/${goos}"/* third-party/
```
Question: Similar question here - should we check exit codes here or keep going even if something failed?

BagToad reviewed (approved):
LGTM! I don't think anything is blocking enough to avoid merge this 👏 

I would like to understand the versioning strategy a bit more, but I don't think that blocks this PR  - more details in my comment on `script/licenses` :)
